src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -34,242 +34,128 @@ private sealed class Parser
+                    ReportDiagnostic(DiagnosticDescriptors.ContextClassesMustBePartial, contextLocation, new string[] { contextTypeSymbol.Name });
+                    return null;
+                }
 
-                        if (_generatedTypes.Count == 0)
-                        {
-                            continue;
-                        }
+                options ??= new JsonSourceGenerationOptionsAttribute();
 
-                        ContextGenerationSpec contextGenSpec = new()
-                        {
-                            ContextType = new(contextTypeSymbol),
-                            GeneratedTypes = _generatedTypes.Values.OrderBy(t => t.TypeRef.FullyQualifiedName).ToImmutableEquatableArray(),
-                            Namespace = contextTypeSymbol.ContainingNamespace.ToDisplayString(),
-                            ContextClassDeclarations = classDeclarationList.ToImmutableEquatableArray(),
-                            DefaultIgnoreCondition = options.DefaultIgnoreCondition,
-                            IgnoreReadOnlyFields = options.IgnoreReadOnlyFields,
-                            IgnoreReadOnlyProperties = options.IgnoreReadOnlyProperties,
-                            IncludeFields = options.IncludeFields,
-                            PropertyNamingPolicy = options.PropertyNamingPolicy,
-                            WriteIndented = options.WriteIndented,
-                        };
-
-                        contextGenSpecList ??= new List<ContextGenerationSpec>();
-                        contextGenSpecList.Add(contextGenSpec);
-
-                        // Clear the caches of generated metadata between the processing of context classes.
-                        _generatedTypes.Clear();
-                        _typesToGenerate.Clear();
-                    }
+                // Enqueue attribute data for spec generation
+                foreach (TypeToGenerate rootSerializableType in rootSerializableTypes)
+                {
+                    EnqueueType(rootSerializableType.Type, rootSerializableType.Mode, rootSerializableType.TypeInfoPropertyName, rootSerializableType.AttributeLocation);
                 }
 
-                if (contextGenSpecList == null)
+                // Walk the transitive type graph generating specs for every encountered type.
+                while (_typesToGenerate.Count > 0)
                 {
-                    return null;
+                    cancellationToken.ThrowIfCancellationRequested();
+                    TypeToGenerate typeToGenerate = _typesToGenerate.Dequeue();
+                    if (!_generatedTypes.ContainsKey(typeToGenerate.Type))
+                    {
+                        TypeGenerationSpec spec = ParseTypeGenerationSpec(typeToGenerate, contextName: contextTypeSymbol.Name, contextLocation, options);
+                        _generatedTypes.Add(typeToGenerate.Type, spec);
```

Comment:
Similar question here.  If we were processing multiple inputs, previously we would have walked all types once, right, even if the same types were used in multiple models?  But now we'll be processing them each once per top-level type?  Or have I misunderstood?

Reply:
Even in the current implementation, the type graphs are being walked independently per context class regardless of overlap. It's probably unavoidable, since each context class could specify different compile-time configuration.